src/pymmcore_widgets/_mda/_mda_widget.py

```diff
@@ -486,13 +489,14 @@ def _update_total_time(self) -> None:
         total_time: float = 0.0
         _per_timepoints: dict[int, float] = {}
         t_per_tp_msg = ""
+        _uses_time = self._uses_time()
 
         for e in self.get_state():
```

Comment:
@ianhi @tlambert03 I think the main issue is here when we call `self.get_state()`. I basically already tried to fix it in this branch (https://github.com/fdrgsp/pymmcore-widgets/tree/update_time_in_thread) using a different thread and it works great on the microscope (no freeze at all). However it gives me the `_run_after_each_test` error in test so I did not open a PR yet.

Reply:
Speeding up `get_state` would be the next big win, and also generically a good thing. However, as it stands the uses_time was actually the most costly part. You can see this from the `line_profiler` breakdown:

```
Timer unit: 1e-09 s

Total time: 18.2211 s
File: /home/ian/Documents/oss/micro/pymmcore-widgets/src/pymmcore_widgets/_mda/_mda_widget.py
Function: _update_total_time at line 468

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   468                                               def _update_total_time(self) -> None:
   469                                                   """Calculate the minimum total acquisition time info."""
   470                                                   # TODO: fix me!!!!!
   471         1      70016.0  70016.0      0.0          if self._mmc.getChannelGroup() and self._mmc.getCurrentConfig(
   472         1       4312.0   4312.0      0.0              self._mmc.getChannelGroup()
   473                                                   ):
   474         1      13291.0  13291.0      0.0              if self.ch_cbox.isChecked() and not self.channel_widget._table.rowCount():
   475                                                           self.time_lbl._total_time_lbl.setText(
   476                                                               "Minimum total acquisition time: 0 sec."
   477                                                           )
   478                                                           return
   479                                           
   480                                                   elif not self.ch_cbox.isChecked() or not self.channel_widget._table.rowCount():
   481                                                       self.time_lbl._total_time_lbl.setText(
   482                                                           "Minimum total acquisition time: 0 sec."
   483                                                       )
   484                                                       return
   485                                           
   486         1        689.0    689.0      0.0          total_time: float = 0.0
   487         1        480.0    480.0      0.0          _per_timepoints: dict[int, float] = {}
   488         1        336.0    336.0      0.0          t_per_tp_msg = ""
   489                                                   # _uses_time = self._uses_time()
   490                                           
   491     36000 4822693525.0 133963.7     26.5          for e in self.get_state():
   492     36000   21320506.0    592.2      0.1              if e.exposure is None:
   493                                                           continue
   494                                           
   495     36000   28594420.0    794.3      0.2              total_time = total_time + (e.exposure / 1000)
   496     36000 13228951745.0 367470.9     72.6              if self._uses_time():
   497     36000   54440471.0   1512.2      0.3                  _t = e.index["t"]
   498     36000   22185299.0    616.3      0.1                  _exp = e.exposure / 1000
   499     36000   42204671.0   1172.4      0.2                  _per_timepoints[_t] = _per_timepoints.get(_t, 0) + _exp
   500                                           
   501         1        443.0    443.0      0.0          if _per_timepoints:
   502         1     320968.0 320968.0      0.0              time_value = self.time_widget.value()
   503                                           
   504         1        263.0    263.0      0.0              intervals = []
   505         1        597.0    597.0      0.0              for phase in time_value["phases"]:  # type: ignore
   506         1       1539.0   1539.0      0.0                  interval = phase["interval"].total_seconds()
   507         1        432.0    432.0      0.0                  intervals.append(interval)
   508         1        452.0    452.0      0.0                  if phase.get("loops") is not None:
   509         1        970.0    970.0      0.0                      total_time = total_time + (phase["loops"] - 1) * interval
   510                                                           else:
   511                                                               total_time = total_time + phase["duration"].total_seconds()
   512                                           
   513                                                       # check if the interval(s) is smaller than the sum of the exposure times
   514         1       7124.0   7124.0      0.0              sum_ch_exp = sum(
   515                                                           (c["exposure"] / 1000)
   516         1      82319.0  82319.0      0.0                  for c in self.channel_widget.value()
   517                                                           if c["exposure"] is not None
   518                                                       )
   519         1        322.0    322.0      0.0              for i in intervals:
   520         1        738.0    738.0      0.0                  if 0 < i < sum_ch_exp:
   521                                                               self.time_widget.setWarningVisible(True)
   522                                                               break
   523                                                           else:
   524         1       5945.0   5945.0      0.0                      self.time_widget.setWarningVisible(False)
   525                                           
   526                                                       # group by time
   527         1     106683.0 106683.0      0.0              _group_by_time: dict[float, list[int]] = {
   528                                                           n: [k for k in _per_timepoints if _per_timepoints[k] == n]
   529         1      21566.0  21566.0      0.0                  for n in set(_per_timepoints.values())
   530                                                       }
   531                                           
   532         1        282.0    282.0      0.0              t_per_tp_msg = "Minimum acquisition time per timepoint: "
   533                                           
   534         1        884.0    884.0      0.0              if len(_group_by_time) == 1:
   535         1        428.0    428.0      0.0                  t_per_tp_msg = (
   536         1        545.0    545.0      0.0                      f"\n{t_per_tp_msg}"
   537         1      35895.0  35895.0      0.0                      f"{fmt_timedelta(timedelta(seconds=_per_timepoints[0]))}"
   538                                                           )
   539                                                       else:
   540                                                           acq_min = timedelta(seconds=min(_per_timepoints.values()))
   541                                                           t_per_tp_msg = (
   542                                                               f"\n{t_per_tp_msg}{fmt_timedelta(acq_min)}"
   543                                                               if self._uses_time()
   544                                                               else ""
   545                                                           )
   546                                                   else:
   547                                                       t_per_tp_msg = ""
   548                                                       self.time_widget.setWarningVisible(False)
   549                                           
   550         1        169.0    169.0      0.0          _min_tot_time = (
   551         1        336.0    336.0      0.0              "Minimum total acquisition time: "
   552         1      12258.0  12258.0      0.0              f"{fmt_timedelta(timedelta(seconds=total_time))}"
   553                                                   )
   554         1       8285.0   8285.0      0.0          self.time_lbl._total_time_lbl.setText(f"{_min_tot_time}{t_per_tp_msg}")

```

Reply:
> I think the main issue is here when we call self.get_state(). I basically already tried to fix it in this branch

I don't think the issue is *calling* `get_state()`, I think the issue is iterating it.  It's an easy trap to fall into, but basically we should pretty much try to avoid ever resorting to iterating over an MDASequence.  Only the final experiment should do that

Reply:
also, from looking at it, uses_time could probably be sped up slightly by changing it from this:

```py
    def _uses_time(self) -> bool:
        """Hacky method to check whether the timebox is selected with any timepoints."""
        has_phases = self.time_widget.value()["phases"]  # type: ignore
        return bool(self.t_cbox.isChecked() and has_phases)
```

to this

```py
    def _uses_time(self) -> bool:
        """Hacky method to check whether the timebox is selected with any timepoints."""
        has_phases = self.time_widget._table.rowCount()
        return bool(self.t_cbox.isChecked() and has_phases)
```

Reply:
also, for what it's worth... i think the whole time estimation thing might be better off as an `estimate_time()` method on MDASequence in useq.

Reply:
The new `_uses_time` is a good microoptimization. That function gets a 205x speed increase from 122 microseconds -> 595 nanoseconds. Sadly we only call it once now so not as much of a speed up as we might've gotten before :(